src/linker/Linker.Analysis/README.md

```diff
@@ -0,0 +1,61 @@
+
+Here, the command-line flags have the following meaning.
+```
+-a console:                    Roots console.dll. The linker will see that it is an application, and root only the entry point (not all public APIs)
+-c link -u link:               Tells the linker to trim all application (-u) and framework (-c) libraries unless otherwise specified.
+-d testapps/console/out:       Tells the linker where to find the dependencies of the application.
+--skip-unresolved:             Continue when encountering unresolved dependencies.
+--ignore-descriptors true:     Don't root types/members specified in embedded .xml descriptors. Use this option to see only what is kept by the linker's analysis.
+--verbose:                     Output extra info to stdout, such as the methods with unanalyzed reflection patterns.
+--used-attrs-only true:        Only keep custom attributes when the attribute type is referenced. (Otherwise this keeps all custom attributes on kept types).
+--out testapps/console/linked: Output the linked app into the specified directory.
+```
+All of the above are standard linker flags. To enable the analyzer:
+```
+--analyze-for-trim:            Output a summary of the analysis to console, and write trimanalysis.json into the linker output (--out) directory.
+```
+
+# How it works
+
+This analyzer maintains a set of "interesting" methods - where "interesting" means either that the linker noticed something special about the method, or we added a (simulated) attribute on it to mark it as interesting explicitly.
+
+The tool finds callers of the interesting method and reports partial "call stacks" to give context on how the linker thought each interesting method was reachable. There are a number of challenges in reporting this information in a useful way, and we will continue iterating on it. For now, it behaves as follows:
+
+The callstacks stop at public APIs or virtual methods. Stopping at virtual methods avoid lots of redundancy in cases where a very common virtual method (like `ToString`) has an override that is unsafe - which would otherwise show up everywhere that `ToString` is called.
+
+The linker already attempts to understand some simple reflection patterns. When it sees a pattern that it doesn't recognize, the containing method is marked as `LinkerUnanalyzed`. Cases where the linker *does* understand a relfection pattern are not yet taken into account. Because we have defined simulated attributes for reflection APIs, most of these will show up as calls to reflection APIs with the `KnownReflection` attribute to indicate that it is not generally safe.
+
+Methods that look "unsafe" but are actually safe can be attributed with a specific attribute to track the feature area, or with `AnnotatedLinkerFriendly` which will prevent the method from showing up in "unsafe" callstacks. The tool may still report callstacks that only end up in `AnnotatedLinkerFriendly` methods (for now), though the plan is for these not to show up in the output.
```

Comment:
We should probably group the reasons into "Known bad" (like `KnownReflection`) and "Marked as safe" (like `AnnotatedLinkerFriendly` or the others). And report it clearly as two separate groups - or maybe hide the second group for now from the output.
I think it's quite confusing that we report basically all callstacks, just with different tags. So even if I go a "fix" a problem by marking something as safe, it will still show up in the output.

Reply:
Agreed - that was the original intention, but I fiddled with the algorithm at some point because I wanted to make sure it was seeing the friendly annotations as expected. I will see about fixing it.